example_allocators/bitmap_allocator/bitmap_alloc.c

```diff
@@ -0,0 +1,202 @@
+ * set to 0). There is one bit per allocatable chunk.
+ *
+ * This is _not_ a clever allocator, since it
+ * does a linear scan of the bitmap to find the
+ * first free chunk, which is expensive!
+ * More efficient scans could be easily incorporated.
+ *
+ * This is _not_ a general-purpose allocator, since
+ * it only allocates chunks of a fixed size. Further,
+ * this size is constrained to be small enough to allow
+ * exact bounds representation in CHERI capabilities.
+ *
+ * This is an initial simple memory allocator test
+ * for CHERI / Capable VMs.
+ * We explore capability alignment,
+ * representable bounds, narrowing operations
+ * and compiler intrinsic support.
+ */
+
+// #include <cheriintrin.h>
+#include <assert.h>
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+
+#include "bitmap_alloc.h"
+
+#define BITS_PER_BYTE 8
+
+char *buffer = NULL;		  /* allocation buffer */
+unsigned char *bitmap = NULL; /* bitmap for the buffer */
+
+int buffer_size = 0;	 /* size of buffer (in bytes) */
+int bitmap_size = 0;	 /* size of bitmap (in bytes) */
+int bytes_per_chunk = 0; /* size of single chunk (in bytes) */
+
+void init_alloc(int num_chunks, int chunk_size)
+{
+	int i = 0;
+
+	/* we need to increase the num_chunks
+	 * so every bit in bitmap will be used
+	 */
+	int adjusted_num_chunks = (num_chunks % BITS_PER_BYTE == 0)
+								  ? num_chunks
+								  : (num_chunks + (BITS_PER_BYTE - (num_chunks % BITS_PER_BYTE)));
+
+	/* we need to increase the chunk_size
+	 * so chunks will be CHERI aligned
+	 * (i.e. 16 bytes for RISC-V 64-bit arch)
+	 */
+	int adjusted_chunk_size =
+		(chunk_size % (sizeof(void *)) == 0)
+			? chunk_size
+			: (chunk_size + (sizeof(void *)) - (chunk_size % (sizeof(void *))));
+
+	/* check this chunk size is small enough
+	 * so we can represent bounds precisely with
+	 * CHERI compressed representation
+	 */
+	// FIXME
+
+	/* request memory for our allocation buffer */
+	char *res = mmap(NULL, adjusted_num_chunks * adjusted_chunk_size, PROT_READ | PROT_WRITE,
+					 MAP_ANON | MAP_PRIVATE, -1, 0);
+	/* request memory for our bitmap */
+	bitmap = (unsigned char *)mmap(NULL, adjusted_num_chunks / BITS_PER_BYTE,
+								   PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0);
+
+	if (res == MAP_FAILED || bitmap == MAP_FAILED)
+	{
+		perror("error in initial mem allocation");
+		exit(-1);
+	}
+
+	/* NB mmap min bounds for capability is 1 page (4K) */
+
+	buffer = res;
+	/* check buffer is aligned */
+	assert((uintptr_t)buffer % sizeof(void *) == 0);
+	/* check bitmap is aligned */
+	assert((uintptr_t)bitmap % sizeof(void *) == 0);
+
+	bytes_per_chunk = adjusted_chunk_size;
+	buffer_size = adjusted_num_chunks * adjusted_chunk_size;
+	bitmap_size = adjusted_num_chunks / BITS_PER_BYTE;
+
+	/* zero bitmap, since all chunks are free initially */
+	for (i = 0; i < bitmap_size; i++)
+	{
+		bitmap[i] = 0;
+	}
+
+	// FIXME - should we use CHERI API to
```

Comment:
We can't set *exact* bounds in all cases but we can set a safe approximate upper limit (i.e. we over-approximate in the worst case), right?

Reply:
This is the older commit within the two commits. This has changed in the second commit.